src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java

```diff
@@ -88,12 +89,29 @@ Optional<Holder<T>> getHolder(T value)
         return Optional.ofNullable(this.holders.get(value));
     }
 
-    Holder<T> getOrCreateHolder(ResourceKey<T> key)
+    DataResult<Holder<T>> getOrCreateHolder(ResourceKey<T> key)
+    {
+        Holder.Reference<T> ref = this.holdersByName.get(key.location());
+        if (ref == null)
+        {
+            if (this.holderLookup != null)
+                return DataResult.error("This registry can't create new holders without value (requested key: " + key + ")");
+
+            if (this.frozen)
+                return DataResult.error("Registry is already frozen (trying to add key " + key + ")");
+
+            ref = Holder.Reference.createStandAlone(this.self, key);
+            this.holdersByName.put(key.location(), ref);
+        }
+        return DataResult.success(ref);
+    }
+
+    Holder<T> getOrCreateHolderOrThrow(ResourceKey<T> key)
     {
         return this.holdersByName.computeIfAbsent(key.location(), k ->
         {
             if (this.holderLookup != null)
-                throw new IllegalStateException("This registry can't create new holders without value");
+                throw new IllegalStateException("This registry can't create new holders without value (requested key: " + key + ")");
             if (this.frozen)
                 throw new IllegalStateException("Registry is already frozen (trying to add key " + k + ")");
             return Holder.Reference.createStandAlone(this.self, key);
```

Comment:
Could we reimplement this method to instead call `getOrCreateHolder` and unwrap the given `DataResult`, turning an error message into an `IllegalStateException` (perhaps using `LambdaExceptionUtils#rethrowConsumer` to bypass the pesky pain of checked exceptions)?

Reply:
I've been advised that the correct lambda exception utility class to suggest is <code>cpw.mods.modlauncher.api.Lam<i>db</i>aExceptionUtils</code> and **not** `cpw.mods.util.LambdaExceptionUtils` -- yes, the one with the misspell of `Lamdba`. Apparently, the one under `cpw.mods.util` is from securejarhandler, which does not export that package, causing an illegal access error during runtime.

Reply:
Right, so, there is a method in DataResult called "getOrThrow" that throws a RuntimeException. Should we use that and have a RuntimeException or have a more specific IllegalStateException?